lib/src/parser/proxy_protocol.rs

```diff
@@ -0,0 +1,226 @@
+use nom::IResult;
+use nom::IResult::*;
+use nom::{be_u8, be_u16, ErrorKind};
+
+use std::net::{Ipv4Addr, Ipv6Addr, SocketAddrV4, SocketAddrV6};
+use std::convert::AsMut;
+use std::convert::From;
+
+use network::protocol::proxy_protocol::header::*;
+
+pub const UNKNOWN_FAMILY: u32 = 1;
+pub const INCORRECT_VERSION: u32 = 2;
+
+const PROTOCOL_SIGNATURE_V2: [u8; 12] = [0x0D, 0x0A, 0x0D, 0x0A, 0x00, 0x0D, 0x0A, 0x51, 0x55, 0x49, 0x54, 0x0A];
+
+//NOTE: Only support the V2 for the moment
+pub fn parse(buf: &[u8]) -> IResult<&[u8], ProxyProtocolHeader> {
+  match tag!(buf, &PROTOCOL_SIGNATURE_V2) {
+    Done(rest, _) => {
+      match parse_v2_header(rest) {
+        Done(rest, header) => Done(rest, ProxyProtocolHeader::V2(header)),
+        Incomplete(i) => Incomplete(i),
+        Error(e) => Error(e),
+      }
+    },
+    Incomplete(i) => Incomplete(i),
+    Error(e) => Error(e),
+  }
+}
+
+//TODO: Refactor this method using cond! parser if it's possible
+fn read_and_verify_ver_and_command(input: &[u8]) -> IResult<&[u8], u8> {
+  match be_u8(input) {
+    Done(rest, ver_and_cmd) => {
+      // Verify if it's the version 2
+      if (ver_and_cmd >> 4) & 0x0f == 0x02 {
+        return Done(rest, ver_and_cmd)
+      }
+      Error(error_code!(ErrorKind::Custom(INCORRECT_VERSION)))
+    },
+    Error(e)      => Error(e),
+    Incomplete(e) => Incomplete(e)
+  }
+}
+
+fn parse_v2_header(input: &[u8]) -> IResult<&[u8], HeaderV2> {
```

Comment:
use `named!` instead of handwritten functions

Reply:
Done